### PR TITLE
refactor: decompose agent internals and harden run/history behavior

### PR DESCRIPTION
**Pull request by ashpreetbedi**
### What
- Decomposes `libs/agno/agno/agent/agent.py` into focused internal modules while keeping the public `Agent` API as a delegation layer.
- Centralizes run option resolution (call-site overrides vs agent defaults) and applies explicit overrides onto an existing `RunContext`.
- Hardens run lifecycle edge cases: cancellation preservation, cleanup of registered runs on failures, continue-run error handling, and history/tooling JSON behavior.
- Adds regression tests covering async-generator delegation, run option resolution, default-tools history JSON, and run/continue-run cancellation/cleanup paths.

### Why
- Improves maintainability/reviewability by splitting the monolithic Agent implementation into coherent units.
- Fixes and locks in correctness for run/history behaviors surfaced during extraction (sync/async parity, cancellation/cleanup, and history tooling).

### Key changes
- `libs/agno/agno/agent/agent.py`: public API delegates into extracted modules.
- `libs/agno/agno/agent/_run.py`: run/stream/continue dispatch + cancellation tracking/cleanup hardening.
- `libs/agno/agno/agent/_run_options.py`: centralized option resolution + metadata/filter merge semantics.
- `libs/agno/agno/agent/_default_tools.py`: history tools return valid JSON (including empty histories).
- `libs/agno/agno/agent/_hooks.py`, `_messages.py`, `_response.py`, `_storage.py`, `_tools.py`, `_init.py`, `_cli.py`, `_telemetry.py`: extracted internal responsibilities.
- Unit tests: `libs/agno/tests/unit/agent/test_async_generator_delegation.py`, `libs/agno/tests/unit/agent/test_run_options.py`, `libs/agno/tests/unit/agent/test_default_tools_json_history.py`, `libs/agno/tests/unit/agent/test_run_regressions.py`.
- Integration tests: update patch targets post-extraction and ensure history-dependent tests set `store_history_messages=True`; skip OpenAI-dependent tests when `OPENAI_API_KEY` is unset.

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [x] Improvement
- [ ] Model update
- [ ] Other:

---

## Checklist

- [x] Code complies with style guidelines
- [x] Ran format/validation scripts (`./scripts/format.sh` and `./scripts/validate.sh`)
- [ ] Self-review completed
- [ ] Documentation updated (comments, docstrings)
- [ ] Examples and guides: Relevant cookbook examples have been included or updated (if applicable)
- [ ] Tested in clean environment
- [x] Tests added/updated (if applicable)

---

## Testing

```bash
source .venv/bin/activate
./scripts/format.sh
./scripts/validate.sh
pytest libs/agno/tests/unit/agent -q
```

Results:
- `./scripts/format.sh` (pass)
- `./scripts/validate.sh` (pass)
- `pytest libs/agno/tests/unit/agent -q` (138 passed)

Optional (requires env/deps):
- `OPENAI_API_KEY=... pytest libs/agno/tests/integration/agent -q`
- `pytest libs/agno/tests/integration/agent/test_agent_run_cancellation.py -q` (some cases also require `ddgs`)

---

## Additional Notes

- OpenAI-backed integration tests in `libs/agno/tests/integration/agent/` are skipped automatically when `OPENAI_API_KEY` is not set.

### CODE REVIEW

**Comment by ashpreetbedi** [comment]
@claude review please

File: libs/agno/agno/agent/_response.py [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+        )
+
+
+async def areason(
+    agent: Agent,
+    run_response: RunOutput,
+    run_messages: RunMessages,
+    run_context: Optional[RunContext] = None,
+    stream_events: Optional[bool] = None,
+) -> Any:
```
Comment (kausmeows) [inline libs/agno/agno/agent/_response.py]: The return type should be `AsyncIterator[RunOutputEvent]` i think

File: libs/agno/agno/agent/_run.py [hunk 2]
```diff
@@ -0,0 +1,3526 @@
...
+
+    # Normalise hook & guardails
+    if not agent._hooks_normalised:
+        if agent.pre_hooks:
+            agent.pre_hooks = normalize_pre_hooks(agent.pre_hooks)  # type: ignore
+        if agent.post_hooks:
+            agent.post_hooks = normalize_post_hooks(agent.post_hooks)  # type: ignore
+        agent._hooks_normalised = True
+
+    session_id, user_id = initialize_session(agent, session_id=session_id, user_id=user_id)
```
Comment (kausmeows) [inline libs/agno/agno/agent/_run.py]: this is being called twice? i see it in line 850 as well.

File: libs/agno/agno/agent/_run.py [hunk 3]
```diff
@@ -0,0 +1,3526 @@
...
+    )
+
+    run_response.model = agent.model.id if agent.model is not None else None
+    run_response.model_provider = agent.model.provider if agent.model is not None else None
+
+    # Start the run metrics timer, to calculate the run duration
+    run_response.metrics = Metrics()
+    run_response.metrics.start_timer()
+
+    yield_run_output = yield_run_output
```
Comment (kausmeows) [inline libs/agno/agno/agent/_run.py]: we can probably clean this self-assignment here

File: libs/agno/agno/agent/_run.py [hunk 4]
```diff
@@ -0,0 +1,3526 @@
...
+        debug_mode: Whether to enable debug mode.
+    """
+    if run_response is None and run_id is None:
+        raise ValueError("Either run_response or run_id must be provided.")
+
+    if run_response is None and (run_id is not None and (session_id is None and agent.session_id is None)):
+        raise ValueError("Session ID is required to continue a run from a run_id.")
+
+    if agent._has_async_db():
+        raise Exception("continue_run() is not supported with an async DB. Please use acontinue_arun() instead.")
```
Comment (kausmeows) [inline libs/agno/agno/agent/_run.py]:
```suggestion
        raise Exception("continue_run() is not supported with an async DB. Please use acontinue_run() instead.")
```

File: libs/agno/agno/agent/_response.py (hunk 1 above)
Comment (ashpreetbedi) [reply libs/agno/agno/agent/_response.py]: Good catch. Updated in commit d6b0cba8a: changed the return type from `Any` to `AsyncIterator[RunOutputEvent]` in `libs/agno/agno/agent/_response.py`.

File: libs/agno/agno/agent/_run.py (hunk 2 above)
Comment (ashpreetbedi) [reply libs/agno/agno/agent/_run.py]: You were right. I removed the duplicate initialize_session(...) invocation in run_dispatch and kept the early initialization path only (commit d6b0cba8a).

File: libs/agno/agno/agent/_run.py (hunk 3 above)
Comment (ashpreetbedi) [reply libs/agno/agno/agent/_run.py]: Agreed. Removed the no-op self-assignment (yield_run_output = yield_run_output) in arun_dispatch in commit d6b0cba8a.

File: libs/agno/agno/agent/_run.py [hunk 5]
```diff
@@ -0,0 +1,3522 @@
...
+    if stream is False:
+        stream_events = False
+
+    if stream_events is None:
+        stream_events = False if agent.stream_events is None else agent.stream_events
+
+    agent.model = cast(Model, agent.model)
+
+    # Get knowledge filters
+    knowledge_filters = knowledge_filters
```
Comment (kausmeows) [inline libs/agno/agno/agent/_run.py]: this self assignment also looks redundant..

File: libs/agno/agno/agent/_run.py [hunk 6]
```diff
@@ -0,0 +1,3522 @@
...
+
+    if stream_events is None:
+        stream_events = False if agent.stream_events is None else agent.stream_events
+
+    # Can't have stream_events if stream is False
+    if stream is False:
+        stream_events = False
+
+    # Get knowledge filters
+    knowledge_filters = knowledge_filters
```
Comment (kausmeows) [inline libs/agno/agno/agent/_run.py]: here as well
